src/cryptonote_protocol/cryptonote_protocol_handler.inl

```diff
@@ -807,9 +809,10 @@ namespace cryptonote
         return 1;
       }
       //to avoid concurrency in core between connections, suspend connections which delivered block later then first one
+      const crypto::hash block_hash = get_block_hash(b);
       if(count == 2)
       {
-        if(m_core.have_block(get_block_hash(b)))
+        if(m_core.have_block(block_hash))
```

Comment:
Yes, this code is meant to do this, but does not do it well. Given the large amount of time that prepare_handle_incoming_blocks takes, it is very likely that all sync threads will go through that first loop, and all see that the second block is not in the blockchain (since the first thread to get the 200 set will likely be stuck in prepare_handle_incoming_blocks by that time). So nothing will be skipped. This is why I put a lock, so prepare_handle_incoming_blocks and adding blocks are done as one, so no secondary syncing thread starts prepare_handle_incoming_blocks needlessly. I'm guessing count==2 is because it "worked better", but I can't see why. Earlying out earlier would solve nothing, due to that large amount of time between receiving and the blockchain getting updated. You'd need some inter thread comms to ask "what do you have ready to add", and that's too complex compared to the lock + dismiss.